src/tsd-metrics-factory.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2016 Groupon.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import os = require("os");
+import path = require("path");
+
+import log4js = require("log4js");
+
+import events = require("events");
+import tsdDef = require("tsdDef");
+import tsdMetrics = require("./tsd-metrics");
+import options = require("./tsd-default-options");
+import log4jsSink = require("./sinks/tsd-query-log-sink");
+import consoleSink = require("./sinks/tsd-console-sink");
+import warningSink = require("./sinks/tsd-warning-sink");
+import hostResolvers = require("./tsd-host-resolvers");
+
+import TsdMetrics = tsdMetrics.TsdMetrics;
+import Options = options.Options;
+
+/**
+ * Factory pattern for creating and configuring instances of Metrics.
+ *
+ * @class
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+export class TsdMetricsFactory implements tsdDef.MetricsFactory {
+    private static defaultHostResolver = new hostResolvers.DefaultHostResolver(os.hostname());
+    private _serviceName:string;
+    private _clusterName:string;
+    private _hostResolver:tsdDef.HostResolver;
+    private _sinks:tsdDef.Sink[];
+    private _logger:log4js.Logger;
+
+    /**
+     * Build an instance of <code>TsdMetricsFactory</code> by providing the full configuration.
+     *
+     * @returns {TsdMetricsFactory}
+     */
+    public static buildInstance(options: {serviceName:string; clusterName:string; hostName?:string; hostResolver?:tsdDef.HostResolver; sinks?:tsdDef.Sink[]; logger?:log4js.Logger}):TsdMetricsFactory {
```

Comment:
Correct: this method takes one argument called `options` which is an object with the specified interface. The interface specifies that both `hostName` and `hostResolver` are optional. If both are provided, `hostResolver` takes priority.